### PR TITLE
[earlgrey, hmac] Add TLT for SHA-2/HMAC multistream

### PR DESCRIPTION
**Pull request by ballifatih**
Reorganize SHA-2/HMAC tests into two:

- One shot calls -> hmac_functest
- Multistream (a.k.a. interleaving streaming) test -> hmac_multistream_functest

also update the test plan and add the second test as TLT, so that it runs on nightlies.

### CODE REVIEW

File: sw/device/tests/crypto/hmac_multistream_functest.c
Comment (moidx) [inline sw/device/tests/crypto/hmac_multistream_functest.c:5]: Remove commented out code.

File: sw/device/tests/crypto/hmac_multistream_functest.c
Comment (moidx) [inline sw/device/tests/crypto/hmac_multistream_functest.c:211]: Would there be any value from randomizing all these loops from a DV coverage perspective? If so we can talk about how to do that in a follow up PR.

**Review by moidx — APPROVED** [review]

File: sw/device/tests/crypto/hmac_multistream_functest.c
Comment (ballifatih) [reply sw/device/tests/crypto/hmac_multistream_functest.c:211]: Yes randomizing the order here would be great. I was not sure if there are available utils regarding randomness and where to get this randomness. I see `sw/device/lib/base/random_order.h` but the implementation looks like a place holder.

At the moment, this test does not configure EDN to save some time, but we can reenable it if it helps with the randomness.

Comment (ballifatih) [reply sw/device/tests/crypto/hmac_multistream_functest.c:210]: An alternative would be to make some bazel changes so that the hardcoded testvector file `sw/device/tests/crypto/testvectors/hmac_testvectors_random.hjson` is re-generated randomly with the following:
```
./sw/device/tests/crypto/hmac_gen_random_testvectors.py 10 sw/device/tests/crypto/testvectors/hmac_testvectors_random.hjson
```
The generation script does not take seed at the moment, but I can add it. It would put less pressure on complication to TLT and we can avoid configuring EDN I think.

In any case, I think we can merge this once CI is clean and discuss how to improve the test. It is also possible to run this test with more vectors on FPGA/nightly.

File: hw/top_earlgrey/data/ip/chip_hmac_testplan.hjson
Comment (gdessouky) [inline hw/top_earlgrey/data/ip/chip_hmac_testplan.hjson:81]: In a future refinement in an upcoming PR, can we randomize the number of update calls, e.g., to range between [1, 4] for the different vectors?

File: hw/top_earlgrey/data/ip/chip_hmac_testplan.hjson
Comment (gdessouky) [inline hw/top_earlgrey/data/ip/chip_hmac_testplan.hjson:83]: minor typo: each stream*

**Review by gdessouky — APPROVED** [review]
thanks @ballifatih, this LGTM (only left a minor typo comment)! 

I think we can discuss more enhancements to the TLT for future PRs but not block this one. It would be great if we can randomize the number of update calls that the 6 vectors end up making, e.g.,  so we end up with a mix where 1 may eventually be one-shot, others are making 2-3 calls, and others more, etc.

**Comment by ballifatih** [comment]
> I think we can discuss more enhancements to the TLT for future PRs but not block this one. It would be great if we can randomize the number of update calls that the 6 vectors end up making, e.g., so we end up with a mix where 1 may eventually be one-shot, others are making 2-3 calls, and others more, etc. WDYT?

As discussed offline with @gdessouky, I created an issue to track the TLT improvements for M5 https://github.com/lowRISC/opentitan/issues/23507

For M4, the focus is to have another PR to update HMAC driver to increase testing coverage.

**Review by martin-velay — APPROVED** [review]
I don't know so much SW, but LGTM

**Pull request merged**
